Miscellaneous build fixes

turboencabulator opened the pull request:
* Make the pkgconfig files (and cmake's equivalent files) mostly identical when generated with either autotools or cmake.
* Build shared libraries by default, and avoid linking transitive dependencies (usually not necessary with shared libs).
* Add parpack.pc and arpackmm.pc.
* Fix a few more files not being installed to their user-requested location.
* Fix Eigen3 detection in cmake.

@fghoussen: Please review.

fghoussen commented on EXAMPLES/MATRIX_MARKET/Makefile.am:
```diff
@@ -6,7 +6,6 @@ EXTRA_DIST = README
 dist_check_SCRIPTS = arpackmm.sh
 dist_check_DATA = An.mtx As.mtx Az.mtx B.mtx Bz.mtx
-SH_LOG_COMPILER = bash
 TESTS = arpackmm.sh
```
@yurivict : wasn't this needed on FreeBSD ? Does `./configure --enable-icb-exmm` runs OK on FreeBSD ?

fghoussen commented on EXAMPLES/MATRIX_MARKET/Makefile.am:
```diff
@@ -1,6 +1,8 @@
 LDADD = $(top_builddir)/SRC/libarpack$(LIBSUFFIX).la
 AM_CPPFLAGS = -I$(top_builddir) -I$(top_srcdir)/ICB $(EIGEN3_CFLAGS)
 
+pkgincludedir = $(includedir)/arpack
```
@turboencabulator : this was explicitly "asked for" at some point (#126). Also both autotools/cmake should behave the same way (not sure what cmake does). I have no religion here but I understand the point from #126.

turboencabulator replied:
This line is for the `pkginclude_HEADERS = arpackSolver.hpp` line below.  Without it, it gets confused on where the file should be installed.
cmake should be equvalent, it has `install(FILES EXAMPLES/MATRIX_MARKET/arpackSolver.hpp DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/arpack)`

fghoussen replied:
OK, sure, missed that this was (only) in `Makefile.am` of `EXAMPLES/MATRIX_MARKET`.

fghoussen commented on configure.ac:
```diff
@@ -245,7 +244,6 @@ AC_SUBST([CMAKE_INSTALL_FULL_LIBDIR], ["$libdir"])  # FIXME: need to recursively
 AC_SUBST([CMAKE_INSTALL_FULL_INCLUDEDIR], ["$includedir"])
 AC_SUBST([BLAS_LIBRARIES], ["$BLAS_LIBS"])  # FIXME: need a filename, not a -l flag
 AC_SUBST([LAPACK_LIBRARIES], ["$LAPACK_LIBS"])
-AC_SUBST([EIGEN3_INCLUDE_DIR], [`echo "$EIGEN3_CFLAGS" | sed -e s/-I//`])
```
Fine to me if this does not break fix for #252: @yurivict : could you confirm this is still OK at your side ?

fghoussen commented on EXAMPLES/MATRIX_MARKET/arpackmm.pc.in:
```diff
@@ -0,0 +1,10 @@
+Version: @PACKAGE_VERSION@
+URL: @PACKAGE_URL@
+Requires: arpack eigen3 >= 3.2
+Cflags: -I${includedir}/arpack
```
To me, pc file is not necessary for `arpackmm`: it is a utility (binary/executable file). AFAIU, pc/cmake files are needed for libraries. @sylvestre may decide or not to add it in the install dir. You will never be able to "link to" `arpackmm`, but, you can use it at the command line prompt !

turboencabulator replied:
I added it since arpackSolver.hpp is getting installed.  If some user needs that header, they can use arpackmm.h to get it.  If nobody needs it, let's not install arpackSolver.hpp and remove arpackmm.pc.

fghoussen replied:
Ah OK, didn't get that this way ! Indeed `arpackSolver.hpp` is installed if you want to compile it from your code : maybe rename `arpackmm.pc` to `arpackSolver.pc` ?

fghoussen replied:
I realize afterwards that : eigen flags are missing and needed to use `arpackSolver`, also `arpack.pc` is enough to use `arpackSolver`(all flags OK and eigen ones too)  => I think you can just kill `arpackmm.pc` !

fghoussen commented:
@turboencabulator : for me, the best way to be clean with pc/cmake files would be to add some dedicated tests (some kind of `installAutotools.sh` and/or `installCMake.sh` tests) that would for instance :  

- install arpack-ng in `/tmp/local`

- set up the env (PKG_CONFIG_PATH, CMAKE_MODULE_PATH)

- copy for instance `arpackmm.cpp` (or some other .f or .cpp program) in a `/tmp/useArpack` dir

- build and run this cpp (or f) from the arpack-ng installed in `/tmp/local` !...

These tests would likely be tricky to get to work... Not sure this is really realistic to hope for such tests to get done !

sylvestre commented on CHANGES:
```diff
@@ -21,6 +20,8 @@ arpack-ng - 3.8.0
 
 [ Kyle Guinn ]
  * [BUG FIX]: fix 'eval: Syntax error: "(" unexpected' error at build time.
+ * Only build shared libraries by default.
```
Please add the argument to explain how to build static again :)

turboencabulator replied:
Done.

turboencabulator commented:
@fghoussen, what is the minimum required version of Eigen3?  Currently everything says >= 3.2.  I have 3.2.7 installed and it fails with a few errors.  I upgraded to 3.3.7 and I don't get the errors, however the compiler eats up all my machine's memory and I have to kill it.  So I'm not able to test --enable-icb-exmm.

turboencabulator commented:
@fghoussen, I like your dedicated tests proposal.  I have no idea how to set up Travis to do it.

fghoussen commented:
> what is the minimum required version of Eigen3?

@turboencabulator : there is no really minimum required version ! :D I mean, I did that besides other things with the eigen version I had at the time : this version was likely 3.2 (with possible warnings ? which would finally break with 3.2.7 ?). For me, the most important thing is that it works with the most current/used and/or latest version (I've just re-checked, on debian, with eigen 3.3.7-2, all compile and run OK like at your side) :  

- compiling `arpackmm` needs to compile all flavors (ss, sn, ds, dn, ...) of all possible `arpackSolver`: lots of templates to instanciate so that compilation is very long and consume lots of memory. autotools uses libtool which seems to get a hard time with templates, but, cmake succeeds to get that done in a long-but-few minutes ! It's even worse with `pyarpack` (based on boost.python and `arpackSolver.hpp`). Don't know how to improve that (compilation time, memory consumption) : if you have any idea, feel free to PR !

- switching minimal version to 3.3.7 could be a good thing, but, this may cause troubles on CI here https://github.com/opencollab/arpack-ng/blob/6dcecc8eef7ef102f02475d7805cddf984fa5e4d/.travis.yml#L114 and elsewhere in `.travis.yml` if latest eigen is not available... And so you would need to compile eigen from source which @sylvestre would likely want to avoid ! https://github.com/opencollab/arpack-ng/blob/6dcecc8eef7ef102f02475d7805cddf984fa5e4d/.travis.yml#L185 Maybe a tradeoff solution could be to change minimal eigen version to 3.3.7 and change `.travis.yml` so that `--enable-exmm` and `-DICBEXMM=ON` are only  tested on latest ubuntu/eoan hoping that eigen prebuilt packages are >= 3.3.7 ?

>  I like your dedicated tests proposal. I have no idea how to set up Travis to do it.

I thought I could commit such tests some time ago, but, this could likely be tricky even it "looks simple"... You likely need to have the test wrapped in `sh` script like https://github.com/opencollab/arpack-ng/blob/6dcecc8eef7ef102f02475d7805cddf984fa5e4d/EXAMPLES/MATRIX_MARKET/arpackmm.sh#L1 and you may have to deal with specific problems both across CI and exotic OS (windows ? FreeBSD ?) not covered by CI : how to know about which temporary dir `/tmp` to use (`/tmp` KO on windows) ? how to create dir `/tmp/useArpack` and copy files in it in a portable (unix, windows) way ? Running bash scripts on FreeBSD or windows may be tricky or fail (according to default interpreter `sh`, `zsh`, ...) ?... I think such tests would be a good thing to get. But sounds like tricky-headache job so I didn't get into it at the time !

sylvestre commented:
@turboencabulator is that ready to be merged? thanks

turboencabulator commented:
@sylvestre Yes it's ready.
@fghoussen Feel free to rename arpackmm.pc if you think a different name would suit it better.

fghoussen commented:
> @fghoussen Feel free to rename arpackmm.pc if you think a different name would suit it better.

Sure ! I'll rename it to `arpackSolver.pc` in a separated PR and switch minimal boost version to 3.3.7 if CI can take it. Pity that `installCMake\Autotools`tricky tests are still in TODO list !

sylvestre commented:
@turboencabulator could you please fix the two conflicts? and I will merge it :)

fghoussen commented:
@turboencabulator : I switched to eigen 3.3. Could you mv `arpackmm.pc` (created here) to `arpackSolver.pc` ?

fghoussen commented:
on master after `./configure --prefix \tmp\local --enable-icb-exmm` I have

```
>> tree /tmp/local/
/tmp/local/
├── include
│   ├── arpack
│   │   ├── arpackdef.h
│   │   ├── arpack.h
│   │   ├── arpack.hpp
│   │   ├── debug_c.h
│   │   ├── debug_c.hpp
│   │   ├── debug.h
│   │   ├── parpack.h
│   │   ├── parpack.hpp
│   │   ├── stat_c.h
│   │   ├── stat_c.hpp
│   │   └── stat.h
│   └── arpack-ng
│       └── arpackSolver.hpp

```

`arpackSolver.hpp` should be in `\tmp\local\include\arpack` : @turboencabulator did you fix that ?

sylvestre commented:
@turboencabulator ping?

turboencabulator commented:
Sorry, I kind of lost interest trying to keep up with the CMake churn.  I'll see if I can get it working again.

turboencabulator commented:
I've rebased those changes and renamed arpackmm.pc to arpackSolver.pc.  The last two commits could use a set of eyes, otherwise everything should be similar to what it was before.

fghoussen commented on arpack-ng-config.cmake.in:
```diff
@@ -18,24 +17,9 @@ set(exec_prefix "@exec_prefix@")
-set_target_properties(ARPACK::ARPACK PROPERTIES INTERFACE_INCLUDE_DIRECTORIES "${arpack_ng_INCLUDE_DIRS}")
-set_target_properties(ARPACK::ARPACK PROPERTIES INTERFACE_LINK_LIBRARIES      "${arpack_ng_LIBRARIES}")
+set_target_properties(ARPACK::ARPACK PROPERTIES INTERFACE_INCLUDE_DIRECTORIES "${includedir}/arpack")
+set_target_properties(ARPACK::ARPACK PROPERTIES INTERFACE_LINK_LIBRARIES      "arpack")
```
Not sure to know if this will work: @turboencabulator , did you test ? https://github.com/opencollab/arpack-ng/pull/259#issuecomment-603829535

turboencabulator replied:
I copied one of the tests and used the example CMake snippet in the comments, and it worked.

From what I've read about CMake, it's discouraged to hard-code paths into these kinds of files.  So it's better in this regard, but maybe not ideal.

fghoussen replied:
I agree. Most important is you tested it OK.

fghoussen commented on arpack-ng-config.cmake.in:
```diff
@@ -18,24 +17,9 @@ set(exec_prefix "@exec_prefix@")
 add_library(PARPACK::PARPACK INTERFACE IMPORTED)
-set_target_properties(PARPACK::PARPACK PROPERTIES INTERFACE_INCLUDE_DIRECTORIES "${arpack_ng_INCLUDE_DIRS}")
-set_target_properties(PARPACK::PARPACK PROPERTIES INTERFACE_LINK_LIBRARIES      "${parpack_ng_LIBRARIES}")
+set_target_properties(PARPACK::PARPACK PROPERTIES INTERFACE_LINK_LIBRARIES    "parpack")
```
Not sure to know if this will work: @turboencabulator , did you test ? https://github.com/opencollab/arpack-ng/pull/259#issuecomment-603829535

sylvestre commented:
@fghoussen would it be possible to run the build in the CI in parallel ? https://travis-ci.org/github/opencollab/arpack-ng/builds/745572876 took more than 9 hours! (this is more than running Firefox on 5 or 6 different archs with almost 100k tests ;)

fghoussen commented:
@sylvestre : I have no idea. AFAIU, stage are parallel but you can't have more. Maybe just the CI is slow. Or you could kill jobs.

fghoussen commented:
CI seems slow anyway (temporary pb at Travis CI ?). Using docker files could help but won't fix this.

sylvestre commented:
Let's land this and see if it sticks :) thanks
